### PR TITLE
avoid copying listarray in unset exec

### LINKED ISSUE #6961: Improve performance of unnest even more
Basically the `Unest` exec plan could be made faster if we reduced some copies. Here is the basic idea in case anyone wants to do that

```
    // Create an array with the unnested values of the list array, given the list
    // array:
    //
    //   [1], null, [2, 3, 4], null, [5, 6]
    //
    // the result array is:
    //
    //   1, null, 2, 3, 4, null, 5, 6
    //
    let unnested_array = unnest_array(list_array)?;
```

This looks very much the same to me as calling `list_array.values()` to get access to the underlying values: https://docs.rs/arrow/latest/arrow/array/struct.GenericListArray.html#method.values

In this case the values array would be more like

```
[1, 2, 3, 4, 5, 6]
```

And the offsets of the list array would be would be like (I think):

```
[0, 1, 1, 3, 3, 6]
```

With a null mask showing the second and fourth element are null

So I was thinking you could calculate the take indices directly from the offsets / nulls without having to copy all the values out of the underlying array

_Originally posted by @alamb in https://github.com/apache/arrow-datafusion/pull/6903#discussion_r1262610832_

### PR DESCRIPTION
**Pull request by smiklos**
# Which issue does this PR close?

Closes #6961.

# Rationale for this change

To avoid unnecessary copying of arrays, improving performance. 

Tbh, I'm not sure these changes make this perform better (can work on some benchmarks), in the end we need the copied/modified array to create the unnested recordbatch and the take indices were already not using the copied data.   

There's also the extra optimization mentioned in the comments regarding the case when `list_array` has no null values. 

# What changes are included in this PR?

UnnestExec is changed such that it uses the take kernel instead of copying values manually + using concat. 

# Are these changes tested?

I've added a unit test to verify calculating the new take indices (specific to list_array and not the other columns in the recordbatch)
The rest should be covered by existing physical plan tests for unnest. 

# Are there any user-facing changes?
Nope

### CODE REVIEW

File: datafusion/core/src/physical_plan/unnest.rs [hunk 1]
```diff
@@ -236,21 +236,21 @@ fn build_batch(
     match list_array.data_type() {
         DataType::List(_) => {
             let list_array = list_array.as_any().downcast_ref::<ListArray>().unwrap();
-            unnest_batch(batch, schema, column, &list_array)
+            unnest_batch(batch, schema, column, &list_array, list_array.values())
```
Comment (smiklos) [inline datafusion/core/src/physical_plan/unnest.rs:239]: `.values` seems to not have a common trait so it needs to be passed down here where we have the concrete types

**Comment by jackwener** [comment]
cc @tustvold @alamb

File: datafusion/core/src/physical_plan/unnest.rs [hunk 2]
```diff
@@ -425,11 +435,16 @@ where
...
+    list_array: &T,
+    list_array_values: &Arc<dyn Array>,
+    list_lengths: &PrimitiveArray<P>,
+) -> Result<Arc<dyn Array + 'static>>
 where
     T: ArrayAccessor<Item = ArrayRef>,
+    P: ArrowPrimitiveType,
 {
-    let elem_type = match list_array.data_type() {
+    let _elem_type = match list_array.data_type() {
```
[the remark this answers is not in the document]
Comment (smiklos) [reply datafusion/core/src/physical_plan/unnest.rs:447]: it does some validation but perhaps it's checked already before the exec runs?

[the remark this answers is not in the document]
Comment (smiklos) [reply datafusion/core/src/physical_plan/unnest.rs]: removed it

**Comment by smiklos** [comment]
I've created a benchmark and ran it against main and this branch and the results look great. This branch seems to perform a lot better. should I commit the benchmark to this pr or make a separate one?

File: datafusion/core/src/physical_plan/unnest.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (jackwener) [reply datafusion/core/src/physical_plan/unnest.rs]: > it does some validation but perhaps it's checked already before the exec runs?

Yes, it was verified in `fn list_lengths<T>(list_array: &T)`

**Comment by smiklos** [comment]
I see one of the tests actually don't pass (anymore?), I'll take a look

**Comment by smiklos** [comment]
Tests should pass now. Got surprised by the values array behaving differently for FixedSizedLists... 
The code could be improved but waiting for feedback first.

**Comment by alamb** [comment]
Thank you @smiklos  -- I plan to review this carefully tomorrow

**Comment by alamb** [comment]
>  I've created a benchmark and ran it against main and this branch and the results look great. This branch seems to perform a lot better. should I commit the benchmark to this pr or make a separate one?


I recommend a separate PR

File: datafusion/core/src/physical_plan/unnest.rs [hunk 3]
```diff
@@ -263,43 +263,27 @@ fn unnest_batch<T>(
...
-    // Depending on the list type the result may be Int32Array or Int64Array.
     let list_lengths = list_lengths(list_array)?;
 
     // Create the indices for the take kernel and then use those indices to create
     // the unnested record batch.
     match list_lengths.data_type() {
         DataType::Int32 => {
             let list_lengths = as_primitive_array::<Int32Type>(&list_lengths)?;
+            let unnested_array =
+                unnest_array(list_array, list_array_values, list_lengths)?;
```
Comment (alamb) [inline datafusion/core/src/physical_plan/unnest.rs:279]: I am still struggling to understand the need to call `unnest_array` and why the take indexes can't be calculated directly against the original values array.

File: datafusion/core/src/physical_plan/unnest.rs (hunk 1 above)
Comment (alamb) [reply datafusion/core/src/physical_plan/unnest.rs:239]: Yes, this makes sense

**Review by alamb — COMMENTED** [review]
Thank you very much @smiklos  -- this looks neat but I don't really understand how it is faster as it still calls `take` twice

It would be great if you could share your benchmark and result with us. If it is faster then I think this PR is good to go.

The test case in dataframe.rs for FixedSizeList is 👨‍🍳 👌 

cc @vincev

**Comment by vincev** [comment]
I run the benchmark I used before, with this PR is about ~10% faster.

This is main (run a few time this is closer to the mean value):

```
./unnest-main
+-----------+
| points    |
+-----------+
| 247563120 |
+-----------+
Elapsed: 6.602
``` 

with this PR:

```
./unnest-pr  
+-----------+
| points    |
+-----------+
| 247563120 |
+-----------+
Elapsed: 6.039
```

This unnest 5M rows into 250M rows.

**Comment by vincev** [comment]
[Here](https://gist.github.com/vincev/bb405d1e7992b8018439b6938cf328cf) are the benchmark and the generator for the data so you can run it and see if you can get the same numbers.

**Comment by smiklos** [comment]
Here is my branch with the benchmark. https://github.com/smiklos/arrow-datafusion/blob/unnest-benchmark/datafusion/core/benches/unnest_query.rs

@vincev it seems you also measure the time it takes to read the parquet file. in my bench I create the values in-memory and saw a lot of speedup (from ~50ms to 10ms or so for larger batches)

File: datafusion/core/src/physical_plan/unnest.rs (hunk 3 above)
Comment (smiklos) [reply datafusion/core/src/physical_plan/unnest.rs:279]: Can you clarify which take indices you mean? 
The ones that were already on main were used to expand the rest of the columns (their values and not the one that is unnested). 

So unnest_array is simply needed as we need to change that column (physically unnest it).
The take indices for the rest of the columns could be calculated based on the values array, it was perhaps prettier to have this intermediate state initially. I can try and get rid of list_sizes all together.

**Comment by smiklos** [comment]
> Thank you very much @smiklos -- this looks neat but I don't really understand how it is faster as it still calls `take` twice
> 
> It would be great if you could share your benchmark and result with us. If it is faster then I think this PR is good to go.
> 
> The test case in dataframe.rs for FixedSizeList is 👨‍🍳 👌
> 
> cc @vincev

It calls take for each column. It may skip calling take for the column being unnested if there are no null values. (that is a special case but can speed up certain queries even more).

There could be another special case if fixed size arrays have only one value per array and no nulls as in that case there's no need for transforming the data. Otherwise for most cases I don't see how we can avoid take

**Comment by smiklos** [comment]
Looking at recent changes needed to unnest, it's best to wait until #7088 is resolved

**Comment by alamb** [comment]
Thank you @smiklos  -- I think we'll get work inspired by this PR in soon. Thank you for pushing us along

**Pull request closed without merging**